fleet/ssh.go

```diff
@@ -3,48 +3,123 @@ package fleet
+
+	// IsActive determines whether the tunnel should be used. This is decided by
+	// the given tunnel configuration. In case this is empty, the SSH tunnel is
+	// not active.
+	IsActive() bool
 }
 
-// NewDialFunc returns an http.Dial function which uses the given ssh tunnel as
-// to connect to the endpoint
-func (t *SSHTunnel) NewDialFunc(config Config) func(string, string) (net.Conn, error) {
-	sshClient, err := ssh.NewSSHClient(t.Username, t.Tunnel, t.getChecker(), true, t.Timeout)
+// NewSSHTunnel creates a new SSH tunnel that is configured with the given
+// settings.
+func NewSSHTunnel(config SSHTunnelConfig) (SSHTunnel, error) {
+	newSSHTunnel := &sshTunnel{
+		SSHTunnelConfig: config,
+		Mutex:           sync.Mutex{},
+	}
+
+	newDialFunc, err := newSSHTunnel.NewDialFunc()
+	if err != nil {
+		return nil, maskAny(err)
+	}
+	newSSHTunnel.HTTPTransport = &http.Transport{
+		Dial: newDialFunc,
+	}
+
+	return newSSHTunnel, nil
+}
+
+type sshTunnel struct {
+	SSHTunnelConfig
+
+	HTTPTransport http.RoundTripper
+	Mutex         sync.Mutex
+}
+
+func (t *sshTunnel) IsActive() bool {
+	return t.Tunnel != ""
+}
+
+// NewDialFunc returns an http.Dial function which uses the given ssh tunnel to
+// to connect to the configured endpoint.
+func (t *sshTunnel) NewDialFunc() (func(string, string) (net.Conn, error), error) {
+	forwardAgent := true
+	sshClient, err := ssh.NewSSHClient(t.Username, t.Tunnel, t.NewHostKeyChecker(), forwardAgent, t.Timeout)
 	if err != nil {
-		fmt.Printf("failed to initialize ssh client: %v\n", maskAny(err))
-		os.Exit(1)
+		return nil, maskAny(err)
 	}
-	tunnelFunc := sshClient.Dial
 
-	if config.Endpoint.Scheme == "unix" || config.Endpoint.Scheme == "file" {
-		target := config.Endpoint.Path
-		tunnelFunc = func(string, string) (net.Conn, error) {
-			cmd := fmt.Sprintf(`fleetctl fd-forward %s`, target)
+	if t.Endpoint.Scheme == "unix" || t.Endpoint.Scheme == "file" {
+		return func(string, string) (net.Conn, error) {
+			cmd := fmt.Sprintf(`fleetctl fd-forward %s`, t.Endpoint.Path)
+			fmt.Printf("cmd: %#v\n", cmd)
 			return ssh.DialCommand(sshClient, cmd)
-		}
+		}, nil
 	}
 
-	return tunnelFunc
+	return sshClient.Dial, nil
 }
 
-// getChecker creates and returns a HostKeyChecker, or nil if any error is encountered
-func (t *SSHTunnel) getChecker() *ssh.HostKeyChecker {
+// NewHostKeyChecker creates a new HostKeyChecker, or nil if any error is
+// encountered.
+func (t *sshTunnel) NewHostKeyChecker() *ssh.HostKeyChecker {
 	if !t.StrictHostKeyChecking {
 		return nil
 	}
 
 	keyFile := ssh.NewHostKeyFile(t.KnownHostsFile)
 	return ssh.NewHostKeyChecker(keyFile)
 }
+
+// RoundTrip provides implementation of http.RoundTripper. This method is
+// simply a synchronized wrapper to the underlying http.Transport.
+func (t *sshTunnel) RoundTrip(req *http.Request) (*http.Response, error) {
+	t.Mutex.Lock()
+	defer t.Mutex.Unlock()
+
+	fmt.Printf("req.URL.Path: %#v\n", req.URL.Path)
+	return t.HTTPTransport.RoundTrip(req)
```

Comment:
This is the discussed approach that actually still not works. What I find strange is that it looks like we do really a lot of requests and I wonder about the amount. 

```
req.URL.Path: "/fleet/v1/units"
cmd: "fleetctl fd-forward /var/run/fleet.sock"
req.URL.Path: "/fleet/v1/units"
req.URL.Path: "/fleet/v1/units"
req.URL.Path: "/fleet/v1/state"
req.URL.Path: "/fleet/v1/state"
req.URL.Path: "/fleet/v1/state"
req.URL.Path: "/fleet/v1/machines"
req.URL.Path: "/fleet/v1/units"
req.URL.Path: "/fleet/v1/units"
req.URL.Path: "/fleet/v1/units"
req.URL.Path: "/fleet/v1/state"
req.URL.Path: "/fleet/v1/state"
req.URL.Path: "/fleet/v1/state"
req.URL.Path: "/fleet/v1/machines"
req.URL.Path: "/fleet/v1/units"
req.URL.Path: "/fleet/v1/units"
req.URL.Path: "/fleet/v1/units"
req.URL.Path: "/fleet/v1/state"
req.URL.Path: "/fleet/v1/state"
req.URL.Path: "/fleet/v1/state"
req.URL.Path: "/fleet/v1/machines"
req.URL.Path: "/fleet/v1/units"
req.URL.Path: "/fleet/v1/units"
req.URL.Path: "/fleet/v1/units"
req.URL.Path: "/fleet/v1/state"
req.URL.Path: "/fleet/v1/state"
req.URL.Path: "/fleet/v1/state"
req.URL.Path: "/fleet/v1/machines"
req.URL.Path: "/fleet/v1/units"
req.URL.Path: "/fleet/v1/units"
cmd: "fleetctl fd-forward /var/run/fleet.sock"
req.URL.Path: "/fleet/v1/units"
cmd: "fleetctl fd-forward /var/run/fleet.sock"
req.URL.Path: "/fleet/v1/units"
```